### PR TITLE
#1294 Fix the problem while doing first call for service in PollConsul provider

### LINKED ISSUE #1294: Ocelot returns 500 from Consul for the first request
I have registered 3 services in Consul. (IdentityService is one of them). While trying to access this service through Ocelot, I had the below exception and Postman response returned 500 Internal server error.

Second request to this service returned 503 Service Unavailable.

Third request was Success and returned 200 OK.

Did I miss something to configure ?

Ocelot Configuration.json

{
      "DownstreamPathTemplate": "/api/v1/identity/register",
      "DownstreamScheme": "https",
      "UseServiceDiscovery": true,
      "ServiceName": "IdentityService",
      "UpstreamPathTemplate": "/v1/identity/register",
      "UpstreamHttpMethod": [ "POST" ],
      "LoadBalancerOptions": {
        "Type": "LeastConnection"
      }
    }
  ],
  "GlobalConfiguration": {
    "RequestIdKey": "OcRequestId",
    "BaseUrl": "https://localhost:44311",
    "ServiceDiscoveryProvider": {
      "Scheme": "http",
      "Host": "localhost",
      "Port": 8500,
      "Type": "Consul"
    }
  }

Exception (500 Reponse):-

Consul.ConsulRequestException: Unexpected response, status code ServiceUnavailable: 

</head><body id=ERR_CONNECT_FAIL>
<div id="titles">
<h1>ERROR</h1>
<h2>The requested URL could not be retrieved</h2>

<p>The following error was encountered while trying to retrieve the URL: <a href="http://localhost:8500/v1/health/service/IdentityService">http://localhost:8500/v1/health/service/IdentityService</a></p>

<blockquote id="error">
<p><b>Connection to ::1 failed.</b></p>

<p id="sysmsg">The system returned: <i>(111) Connection refused</i></p>

<p>The remote host or network may be down. Please try the request again.</p>

<p>Your cache administrator is <a href="mailto:webmaster?subject=CacheErrorInfo%20-%20ERR_CONNECT_FAIL&amp;body=CacheHost%3A%20moatcmn-deb1%0D%0AErrPage%3A%20ERR_CONNECT_FAIL%0D%0AErr%3A%20(111)%20Connection%20refused%0D%0ATimeStamp%3A%20Mon,%2020%20Jul%202020%2006%3A16%3A02%20GMT%0D%0A%0D%0AClientIP%3A%20172.21.206.228%0D%0AServerIP%3A%20localhost%0D%0A%0D%0AHTTP%20Request%3A%0D%0AGET%20%2Fv1%2Fhealth%2Fservice%2FIdentityService%20HTTP%2F1.1%0AAccept%3A%20application%2Fjson%0D%0AKeep-Alive%3A%20true%0D%0AHost%3A%20localhost%3A8500%0D%0A%0D%0A%0D%0A">webmaster</a>.</p>

<br>
</div>

<hr>
<div id="footer">
<p>Generated Mon, 20 Jul 2020 06:16:02 GMT by moatcmn-deb1 (squid/3.5.23)</p>

</div>
</body></html>

   at Consul.GetRequest`1.Execute(CancellationToken ct)
   at Ocelot.Provider.Consul.Consul.Get()
   at Ocelot.LoadBalancer.LoadBalancers.LoadBalancerFactory.<>c__DisplayClass2_0.<<Get>b__1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Ocelot.LoadBalancer.LoadBalancers.LeastConnection.Lease(DownstreamContext downstreamContext)
   at Ocelot.LoadBalancer.Middleware.LoadBalancingMiddleware.Invoke(DownstreamContext context)
   at Ocelot.QueryStrings.Middleware.ClaimsToQueryStringMiddleware.Invoke(DownstreamContext context)
   at Ocelot.Headers.Middleware.ClaimsToHeadersMiddleware.Invoke(DownstreamContext context)
   at Ocelot.Authorisation.Middleware.AuthorisationMiddleware.Invoke(DownstreamContext context)
   at Ocelot.Claims.Middleware.ClaimsToClaimsMiddleware.Invoke(DownstreamContext context)
   at Ocelot.Authentication.Middleware.AuthenticationMiddleware.Invoke(DownstreamContext context)
   at Ocelot.RequestId.Middleware.ReRouteRequestIdMiddleware.Invoke(DownstreamContext context)
   at Ocelot.RateLimit.Middleware.ClientRateLimitMiddleware.Invoke(DownstreamContext context)
   at Ocelot.Request.Middleware.DownstreamRequestInitialiserMiddleware.Invoke(DownstreamContext context)
   at Ocelot.Headers.Middleware.HttpHeadersTransformationMiddleware.Invoke(DownstreamContext context)
   at Ocelot.Security.Middleware.SecurityMiddleware.Invoke(DownstreamContext context)
   at Ocelot.Middleware.Multiplexer.Multiplexer.Fire(DownstreamContext context, OcelotRequestDelegate next)
   at Ocelot.Middleware.Multiplexer.Multiplexer.Multiplex(DownstreamContext context, ReRoute reRoute, OcelotRequestDelegate next)
   at Ocelot.DownstreamRouteFinder.Middleware.DownstreamRouteFinderMiddleware.Invoke(DownstreamContext context)
   at Ocelot.Responder.Middleware.ResponderMiddleware.Invoke(DownstreamContext context)
   at Ocelot.Middleware.Pipeline.MapWhenMiddleware.Invoke(DownstreamContext context)
   at Ocelot.Errors.Middleware.ExceptionHandlerMiddleware.Invoke(DownstreamContext context) 
2020-07-20 11:46:01.3982||DEBUG|Ocelot.Errors.Middleware.ExceptionHandlerMiddleware|requestId: 0HM1CCO82S6AR:00000001, previousRequestId: no previous request id, message: ocelot pipeline finished 

Ocelot : 13.5.2
Net Core: 2.2

Any help is appreciated..

### PR DESCRIPTION
**Pull request by przemyslawandruszewski**
## Fixes #1294 
- #1294 

Fix for `PollConsul` provider for the first request to any service while API gateway uses `PollConsul` provider.

## Proposed Changes
  - Fix the problem of first request before timer will start fetching services from Consul

### CODE REVIEW

**Comment by przemyslawandruszewski** [comment]
@TomPallister  could you have a look? :)

**Comment by przemyslawandruszewski** [comment]
Hi @TomPallister , could you please have a look? :)

**Comment by raman-m** [comment]
> I could have a look...

Hi Przemyslaw!
Could you resolve merge conflicts first, please?

Don't forget to update your feature branch by latest top-commits from **base:develop** branch!

File: test/Ocelot.UnitTests/Consul/ProviderFactoryTests.cs
Comment (raman-m) [inline test/Ocelot.UnitTests/Consul/ProviderFactoryTests.cs:1]: Why did you include this file? I see zero changes in this file. Are the some changes in this file?
Please note, styling and formatting should not be applied. You have to show real changes only!

**Comment by raman-m** [comment]
Przemyslaw!
There are a lot of merge conflicts!
I guess current feature branch is outdated.
It is better to backup your changes to some folder. And re-fork the Ocelot repo. SO, you will have top commits from head repo. And create new feature branch from develop. Apply your code changes. Make new PR please!

Or go to my forked [Ocelot repo](/raman-m/Ocelot). And it says: "This branch is up to date with ThreeMammals/Ocelot:develop."
I can add you as Contributor. You will be able to create feature branches.

Do you have an intention to have some pair programming?

**Comment by raman-m** [comment]
Przemyslaw,
I thought you needed my help to resolve merge conflicts. 😉 
The [feature branch](/przemyslawandruszewski/Ocelot/tree/bugfix/fix-pollconsul-first-request-problem) has been [rebased](#event-10307090287) onto [ThreeMammals:develop](../tree/develop) successfully!
I've additionally linked to the issue from [your past comment](#issuecomment-1026846724).

Good luck during code review!

File: src/Ocelot.Provider.Consul/PollConsul.cs [hunk 1]
```diff
@@ -41,13 +51,22 @@ public void Dispose()
...
+        }
+
+        await Poll();
+
+        return _cache.Get<List<Service>>(_serviceCacheKey);
     }
 
     private async Task Poll()
     {
-        _services = await _consulServiceDiscoveryProvider.Get();
```
Comment (RaynaldM) [inline src/Ocelot.Provider.Consul/PollConsul.cs:51]: The "Poll" method mixes 2 things: getting services and caching. I think that respecting the principle of single responsibility, and putting caching in the main "Get" method, would make the code more readable.
Going further, is the "Poll" method really useful?

**Review by RaynaldM — APPROVED** [review]

**Comment by ggnaegi** [comment]
@raman-m This should be addressed in https://github.com/ThreeMammals/Ocelot/pull/1670, no?

**Comment by raman-m** [comment]
>@ggnaegi commented [on Sep 18](#issuecomment-1723452367)

Yes, it is addressed, but this PR has extra enhancement of [injecting IMemoryCache object](1562/files#diff-0dc582946e7840a5fc9034ea15827ccfeeba0d17e7007bc336467aee00a5a9f6R27) into `PollConsul` constructor! 😉

File: src/Ocelot.Provider.Consul/PollConsul.cs (hunk 1 above)
Comment (raman-m) [reply src/Ocelot.Provider.Consul/PollConsul.cs:51]: Ray, is caching a good functionality at all in the provider?

**Comment by ggnaegi** [comment]
> > @ggnaegi commented [on Sep 18](#issuecomment-1723452367)
> 
> Yes, it is addressed, but this PR has extra enhancement of [injecting IMemoryCache object](1562/files#diff-0dc582946e7840a5fc9034ea15827ccfeeba0d17e7007bc336467aee00a5a9f6R27) into `PollConsul` constructor! 😉

Like @RaynaldM I think, we are mixing concerns here...

**Comment by raman-m** [comment]
@przemyslawandruszewski 
In future we can inject caching interface, but it should not be memory cache, otherwise Ocelot will duplicate all collections from Consul service... I understand, you've tried to improve the performance, but the root cause is 
Sorry, cannot accept it nowadays. Cache must be more abstract. And we cannot add this caching forcibly, so configuration property is required.
If you have an intention to contribute, please return back with reopen or message.

Again, Thanks you for the PR!

**Comment by raman-m** [comment]
#1294 will be fixed by #1670

**Pull request closed without merging**